### ISSUE TITLE
Static analysis proof-of-concept

### ISSUE THREAD

**Issue by pmjones** [qiqphp/qiq#13]
@koriym @harikt and anyone else interested:

~~This branch <https://github.com/qiqphp/qiq/tree/1.x-stan-poc> is a proof-of-concept for enabling static analysis on Qiq template files.~~ **Here is a more limited/restricted POC branch, with only the Container and assigned-variable differences. <https://github.com/qiqphp/qiq/tree/1.x-stan-poc-limited>**

~~The diff is at https://github.com/qiqphp/qiq/compare/1.x-stan-poc?expand=1~~ **Limited/restricted diff here <https://github.com/qiqphp/qiq/compare/1.x-stan-poc-limited?expand=1>**

It solves a couple of problems raised by @koriym and others, namely that of helper-object creation through containers, and of template analysis. It is a substantial BC break and will warrant a 2.x release if we go with it.

tl;dr:

- HelperLocator is now a ContainerInterface implementation
- You don't register helpers, you create traits for them, and `use` those traits inside a Template class of your own
- Assigned variables are `$foo` and not `$this->foo`
- Use docblocks in the template file to enable static analysis

Look to the `README-STAN.md` file for detailed information.

Please let me know what you think, if you have alternative implementation ideas, if it's dumb and everything is fine as it is now, etc.

Thanks, and Happy New Year!

**Comment by harikt** [comment]
Thank you @pmjones for your good work. The diff is very big. So wasn't able to go through all the details.

1. First impression was I didn't liked the trait usage.

Reason : 
1. You want to create another class and cannot start on the fly.

2 . Second impression.

I think that make sense because you only want to create the class once for the project or can create another package which extends this.

**Comment by pmjones** [comment]
Hi @harikt thanks for this feedback -- very valuable!

> The diff is very big.

It *is* big; I should have done a better job of isolating the changes. I will see if I can go back and work up a smaller diff that highlights just the substantial elements.

> I didn't liked the trait usage. ... You want to create another class and cannot start on the fly.

I don't especially like it either, for a similar reason. I can rationalize it as "this is not much different than registering it with a service locator", but it's not quite the same. As a side note, there doesn't *have* to be a trait, you can create a method directly on your extended Template class instead, and that is less work.

> you only want to create the class once for the project 

You mean the helper class, or the extended Template class?

**Comment by pmjones** [comment]
@harikt @koriym I have updated the summary above with links to a more restricted/limited branch and diff at https://github.com/qiqphp/qiq/tree/1.x-stan-poc-limited -- sorry for the noise.

**Comment by harikt** [comment]
> You mean the helper class, or the extended Template class?

Extending the template class.

**Comment by pmjones** [comment]
@harikt @koriym I have found an interesting PHPStan annotation, `@mixin`:

https://phpstan.org/writing-php-code/phpdocs-basics#mixins

That looks very powerful, and might help this static-analysis initiative, in that we can put helper methods into a separate helper object, instead of on an extended template class.

**Comment by pmjones** [comment]
@harikt @koriym the `@mixin` tag was a bust, but it led me to intersection typehints, which led to this revamped POC.

If you would, please take a look at the updated https://github.com/qiqphp/qiq/blob/1.x-stan-poc/README-STAN.md file, it has the new summary. In short:

- No need to extend the _Template_ object
- Instead, extend and inject a _Helpers_ object
- A helper is just method on the extended _Helpers_ object
- Traits and objects not required for helpers, but can be useful if needed
- Template files need an intersection typehint to enable static analysis
- Extending the Template class with a `@mixin` tag can make the template file typehint shorter

I think this is a substantial improvement over the prior POC revision. Please let me know what you think!

**Comment by harikt** [comment]
I am sorry for being late. Busy days. But did peep into it. Over all it looks good to me. I will look more closely when I get more time and give you feedback if I change my mind.

**Comment by koriym** [comment]
I think it's a good idea. I would have liked to see the "no registration required, just put a helper file" that I suggested earlier, but it is practical enough this way.

**Comment by pmjones** [comment]
Closing this in favor of #19 -- thanks, all!
